File: AI_UoB/programs/data_structures_for_ai/2_linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def insertAtEnd(self, value):
        newNode = Node(value)

        if self.head is None:
            self.head = newNode
            return

        temp = self.head

        while temp.next is not None:
            temp = temp.next

        temp.next = newNode

    def insertAtStart(self, value):
        newNode = Node(value)
        newNode.next = self.head
        self.head = newNode

    def insertAtPosition(self, index, value):
        if index == 0:
            self.insertAtStart(value)
            return

        newNode = Node(value)
        temp = self.head

        for i in range(index - 1):
            temp = temp.next

        newNode.next = temp.next
        temp.next = newNode

    def display(self):
        temp = self.head
        result = "HEAD -> "

        while temp is not None:
            result += str(temp.data) + " -> "
            temp = temp.next

        result += "NULL"
        print(result)

    def readAt(self, index):
        temp = self.head

        for i in range(index):
            temp = temp.next

        return temp.data

    def update(self, index, value):
        temp = self.head

        for i in range(index):
            temp = temp.next

        temp.data = value

    def deleteAt(self, index):
        if index == 0:
            self.head = self.head.next
            return

        temp = self.head

        for i in range(index - 1):
            temp = temp.next

        temp.next = temp.next.next


if __name__ == "__main__":
    ll = LinkedList()

    ll.insertAtEnd(10)
    ll.insertAtEnd(20)
    ll.insertAtEnd(30)
    ll.insertAtEnd(40)
    ll.insertAtEnd(50)

    ll.display()

    ll.insertAtStart(5)
    ll.display()

    ll.insertAtPosition(3, 99)
    ll.display()

    ll.update(3, 77)
    ll.display()

    ll.deleteAt(3)
    ll.display()

    print(ll.readAt(2))

File: AI_UoB/programs/data_structures_for_ai/2_linkedlist.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* head = NULL;

struct Node* createNode(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->next = NULL;
    return newNode;
}

void insertAtEnd(int value) {
    struct Node* newNode = createNode(value);

    if (head == NULL) {
        head = newNode;
        return;
    }

    struct Node* temp = head;

    while (temp->next != NULL)
        temp = temp->next;

    temp->next = newNode;
}

void insertAtStart(int value) {
    struct Node* newNode = createNode(value);
    newNode->next = head;
    head = newNode;
}

void insertAtPosition(int index, int value) {
    if (index == 0) {
        insertAtStart(value);
        return;
    }

    struct Node* newNode = createNode(value);
    struct Node* temp = head;

    for (int i = 0; i < index - 1; i++)
        temp = temp->next;

    newNode->next = temp->next;
    temp->next = newNode;
}

void display() {
    struct Node* temp = head;

    printf("HEAD -> ");

    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }

    printf("NULL\n");
}

int readAt(int index) {
    struct Node* temp = head;

    for (int i = 0; i < index; i++)
        temp = temp->next;

    return temp->data;
}

void update(int index, int value) {
    struct Node* temp = head;

    for (int i = 0; i < index; i++)
        temp = temp->next;

    temp->data = value;
}

void deleteAt(int index) {
    if (index == 0) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
        return;
    }

    struct Node* temp = head;

    for (int i = 0; i < index - 1; i++)
        temp = temp->next;

    struct Node* toDelete = temp->next;
    temp->next = toDelete->next;

    free(toDelete);
}

int main() {
    insertAtEnd(10);
    insertAtEnd(20);
    insertAtEnd(30);
    insertAtEnd(40);
    insertAtEnd(50);

    display();

    insertAtStart(5);
    display();

    insertAtPosition(3, 99);
    display();

    update(3, 77);
    display();

    deleteAt(3);
    display();

    printf("%d\n", readAt(2));

    return 0;
}